### ISSUE TITLE
Client segfault when "backup_script_pre_arg" is set

### ISSUE THREAD

**Issue by jleroy** [grke/burp#318]
I'm trying to execute a pre-backup script with the following options: 

```
backup_script_pre = /etc/burp/pre_backup_script
backup_script_pre_arg = mariadb
backup_script_pre_arg = database
```

However, the burp client segfault before launching the script: 

```
# burp -a b
2015-08-09 01:05:27: burp[16970] before client
2015-08-09 01:05:27: burp[16970] begin client
2015-08-09 01:05:27: burp[16970] WARNING: Client 'XXXXXX' version '1.4.40' does not match server version '1.3.48'. An upgrade is recommended.
2015-08-09 01:05:27: burp[16970] auth ok
2015-08-09 01:05:27: burp[16970] Server version: 1.3.48
2015-08-09 01:05:27: burp[16970] nocsr ok
2015-08-09 01:05:27: burp[16970] SSL is using cipher: DHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=DH       Au=RSA  Enc=AESGCM(256) Mac=AEAD
2015-08-09 01:05:27: burp[16970] Server is setting includes/excludes.
2015-08-09 01:05:27: burp[16970] Server is overriding the configuration
2015-08-09 01:05:27: burp[16970] with the following settings:
2015-08-09 01:05:27: burp[16970] include = /home
2015-08-09 01:05:27: burp[16970] include = /root
2015-08-09 01:05:27: burp[16970] include = /var/backups.local
2015-08-09 01:05:27: burp[16970] include = /var/www
2015-08-09 01:05:27: burp[16970] exclude = /var/www/XXXXXX/public/images
2015-08-09 01:05:27: burp[16970] exclude_ext = lock
2015-08-09 01:05:27: burp[16970] exclude_ext = tmp
2015-08-09 01:05:27: burp[16970] exclude_regex = (.+)/tmp/(.+)
2015-08-09 01:05:27: burp[16970] exclude_regex = (.+)/cache/(.+)
2015-08-09 01:05:27: burp[16970] exclude_comp = bz2
2015-08-09 01:05:27: burp[16970] exclude_comp = gz
2015-08-09 01:05:27: burp[16970] cross_all_filesystems = 0
2015-08-09 01:05:27: burp[16970] split_vss = 0
2015-08-09 01:05:27: burp[16970] strip_vss = 0
2015-08-09 01:05:27: burp[16970] read_all_fifos = 0
2015-08-09 01:05:27: burp[16970] min_file_size = 0
2015-08-09 01:05:27: burp[16970] max_file_size = 0
2015-08-09 01:05:27: burp[16970] Compression level: 9
Segmentation fault
```

If I comment the `backup_script_pre_arg` options, everything works fine: 

```
2015-08-09 01:05:00: burp[16933] before client
2015-08-09 01:05:00: burp[16933] begin client
2015-08-09 01:05:00: burp[16933] WARNING: Client 'XXXXXX' version '1.4.40' does not match server version '1.3.48'. An upgrade is recommended.
2015-08-09 01:05:00: burp[16933] auth ok
2015-08-09 01:05:00: burp[16933] Server version: 1.3.48
2015-08-09 01:05:00: burp[16933] nocsr ok
2015-08-09 01:05:00: burp[16933] SSL is using cipher: DHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=DH       Au=RSA  Enc=AESGCM(256) Mac=AEAD
2015-08-09 01:05:00: burp[16933] Server is setting includes/excludes.
2015-08-09 01:05:00: burp[16933] Server is overriding the configuration
2015-08-09 01:05:00: burp[16933] with the following settings:
2015-08-09 01:05:00: burp[16933] include = /home
2015-08-09 01:05:00: burp[16933] include = /root
2015-08-09 01:05:00: burp[16933] include = /var/backups.local
2015-08-09 01:05:00: burp[16933] include = /var/www
2015-08-09 01:05:00: burp[16933] exclude = /var/www/XXXXXX/public/images
2015-08-09 01:05:00: burp[16933] exclude_ext = lock
2015-08-09 01:05:00: burp[16933] exclude_ext = tmp
2015-08-09 01:05:00: burp[16933] exclude_regex = (.+)/tmp/(.+)
2015-08-09 01:05:00: burp[16933] exclude_regex = (.+)/cache/(.+)
2015-08-09 01:05:00: burp[16933] exclude_comp = bz2
2015-08-09 01:05:00: burp[16933] exclude_comp = gz
2015-08-09 01:05:00: burp[16933] cross_all_filesystems = 0
2015-08-09 01:05:00: burp[16933] split_vss = 0
2015-08-09 01:05:00: burp[16933] strip_vss = 0
2015-08-09 01:05:00: burp[16933] read_all_fifos = 0
2015-08-09 01:05:00: burp[16933] min_file_size = 0
2015-08-09 01:05:00: burp[16933] max_file_size = 0
2015-08-09 01:05:02: burp[16933] Compression level: 9
2015-08-09 01:05:02: burp[16933] Running pre-backup script: pre reserved2 reserved3 reserved4 reserved5
2015-08-09 01:05:02: burp[16933] 
2015-08-09 01:05:02: burp[16933] Error: invalid database type ("").
2015-08-09 01:05:02: burp[16933] /etc/burp/pre_backup_script returned: 1
2015-08-09 01:05:02: burp[16933] WARNING: /etc/burp/pre_backup_script returned: 1
2015-08-09 01:05:02: burp[16933] error in backup
2015-08-09 01:05:02: burp[16933] after client
```

Server-side log: 

```
# cat working/log 
2015-08-09 01:07:47: burp[9164] Client version: 1.4.40
2015-08-09 01:07:47: burp[9164] WARNING: Client 'XXXXXX' version '1.4.40' does not match server version '1.3.48'. An upgrade is recommended.
2015-08-09 01:07:47: burp[9164] Begin phase1 (file system scan)
2015-08-09 01:07:47: burp[9164] Got SSL_ERROR_SYSCALL in read, errno=0 (Success)
2015-08-09 01:07:47: burp[9164] SSL read problem
2015-08-09 01:07:47: burp[9164] End phase1 (file system scan)
2015-08-09 01:07:47: burp[9164] error in phase 1
```

GDB backtrace: 

```
#0  run_script_to_buf (args=args@entry=0x7fffffffd950, userargs=<optimized out>, userargc=2, cntr=cntr@entry=0x7fffffffded0, do_wait=do_wait@entry=1, logfunc=logfunc@entry=1, logbuf=0x0)
    at handy.c:1133
        a = 0
        serr = 0x0
        s = 0
        l = 6
        p = <optimized out>
        sout = 0x0
        cmd = {0x6415c0 "/etc/burp/pre_backup_script", 0x431d68 "pre", 0x430b96 "reserved2", 0x430ba0 "reserved3", 0x430baa "reserved4", 0x430bb4 "reserved5", 0x0 <repeats 58 times>}
#1  0x000000000041c74b in run_script (args=args@entry=0x7fffffffd950, userargs=<optimized out>, userargc=<optimized out>, cntr=cntr@entry=0x7fffffffded0, do_wait=do_wait@entry=1, 
    logfunc=logfunc@entry=1) at handy.c:1191
No locals.
#2  0x000000000040fdd7 in backup_wrapper (act=ACTION_BACKUP, phase1str=<optimized out>, incexc=<optimized out>, name_max=0, p1cntr=0x7fffffffded0, cntr=0x7fffffffdcc0, conf=0x7fffffffe1a0, 
    resume=0) at client.c:224
        a = 7
        args = {0x6415c0 "/etc/burp/pre_backup_script", 0x431d68 "pre", 0x430b96 "reserved2", 0x430ba0 "reserved3", 0x430baa "reserved4", 0x430bb4 "reserved5", 0x0, 
          0x63 <error: Cannot access memory at address 0x63>, 0x0, 0x677cd0 "ok:9", 0x0, 0x406424 <async_read_expect(char, char const*)+132> "\211\330H\203\304 []A\\A]A^\303\017\037D"}
        ret = <optimized out>
#3  0x00000000004109de in do_client (conf=0x7fffffffe1a0, act=ACTION_BACKUP, vss_restore=2, vss_restore@entry=1, json=0) at client.c:544
        ctx = 0x65b980
        cntr = {gtotal = 0, gtotal_same = 0, gtotal_changed = 0, gtotal_deleted = 0, total = 0, total_same = 0, total_changed = 0, total_deleted = 0, file = 0, file_same = 0, 
          file_changed = 0, file_deleted = 0, enc = 0, enc_same = 0, enc_changed = 0, enc_deleted = 0, meta = 0, meta_same = 0, meta_changed = 0, meta_deleted = 0, encmeta = 0, 
          encmeta_same = 0, encmeta_changed = 0, encmeta_deleted = 0, dir = 0, dir_same = 0, dir_changed = 0, dir_deleted = 0, slink = 0, slink_same = 0, slink_changed = 0, 
          slink_deleted = 0, hlink = 0, hlink_same = 0, hlink_changed = 0, hlink_deleted = 0, special = 0, special_same = 0, special_changed = 0, special_deleted = 0, efs = 0, 
          efs_same = 0, efs_changed = 0, efs_deleted = 0, vss = 0, vss_same = 0, vss_changed = 0, vss_deleted = 0, encvss = 0, encvss_same = 0, encvss_changed = 0, encvss_deleted = 0, 
          vss_t = 0, vss_t_same = 0, vss_t_changed = 0, vss_t_deleted = 0, encvss_t = 0, encvss_t_same = 0, encvss_t_changed = 0, encvss_t_deleted = 0, warning = 0, byte = 0, recvbyte = 0, 
          sentbyte = 0, start = 1439076203}
        name_max = 0
        rfd = -1
        sbio = 0x6
        p1cntr = {gtotal = 0, gtotal_same = 0, gtotal_changed = 0, gtotal_deleted = 0, total = 0, total_same = 0, total_changed = 0, total_deleted = 0, file = 0, file_same = 0, 
          file_changed = 0, file_deleted = 0, enc = 0, enc_same = 0, enc_changed = 0, enc_deleted = 0, meta = 0, meta_same = 0, meta_changed = 0, meta_deleted = 0, encmeta = 0, 
          encmeta_same = 0, encmeta_changed = 0, encmeta_deleted = 0, dir = 0, dir_same = 0, dir_changed = 0, dir_deleted = 0, slink = 0, slink_same = 0, slink_changed = 0, 
          slink_deleted = 0, hlink = 0, hlink_same = 0, hlink_changed = 0, hlink_deleted = 0, special = 0, special_same = 0, special_changed = 0, special_deleted = 0, efs = 0, 
          efs_same = 0, efs_changed = 0, efs_deleted = 0, vss = 0, vss_same = 0, vss_changed = 0, vss_deleted = 0, encvss = 0, encvss_same = 0, encvss_changed = 0, encvss_deleted = 0, 
          vss_t = 0, vss_t_same = 0, vss_t_changed = 0, vss_t_deleted = 0, encvss_t = 0, encvss_t_same = 0, encvss_t_changed = 0, encvss_t_deleted = 0, warning = 1, byte = 0, recvbyte = 0, 
          sentbyte = 0, start = 1439076203}
        incexc = 0x677f30 "include = /home\ninclude = /root\ninclude = /var/backups.local\ninclude = /var/www\nexclude = /var/www/www.dev.collectorsquare.com/public/images\nexclude_ext = lock\nexclude_ext = tmp\nexclude_regex = (.+)/t"...
        server_version = 0x6726d0 "1.3.48"
        ssl = 0x0
#4  0x0000000000410a4a in client (conf=conf@entry=0x7fffffffe1a0, act=act@entry=ACTION_BACKUP, vss_restore=vss_restore@entry=1, json=json@entry=0) at client.c:635
        ret = 0
#5  0x00000000004052ae in main (argc=<optimized out>, argv=<optimized out>) at prog.c:373
        forking = 1
        backup = 0x0
        orig_client = <optimized out>
        generate_ca_only = 0
        __FUNCTION__ = "main"
        option = <optimized out>
        browsedir = 0x0
        configfile = 0x640540 <get_config_path()::path> "/etc/burp/burp.conf"
        json = 0
        act = ACTION_BACKUP
        restoreprefix = 0x0
        vss_restore = 1
        test_conf = 0
---Type <return> to continue, or q <return> to quit---
        ret = 0
        strip = 0
        forceoverwrite = 0
        regex = 0x0
        sclient = 0x0
        daemon = 1
        gotlock = 1
        conf = {configfile = 0x641010 "/etc/burp/burp.conf", port = 0x641270 "4971", status_port = 0x0, mode = MODE_CLIENT, lockfile = 0x641300 "/var/run/burp.client.pid", 
          log_to_syslog = 1, log_to_stdout = 1, progress_counter = 1, ssl_cert_ca = 0x641370 "/etc/burp/ssl_cert_ca.pem", ssl_cert = 0x6413a0 "/etc/burp/ssl_cert-client.pem", 
          ssl_key = 0x6413d0 "/etc/burp/ssl_cert-client.key", ssl_key_password = 0x6414a0 "", ssl_peer_cn = 0x641480 "backup1.inikup.com", 
          ssl_ciphers = 0x641400 "ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AES:RSA+3DES:!ADH:!AECDH:!MD5:!DSS", user = 0x0, group = 0x0, 
          ratelimit = 0, network_timeout = 7200, client_is_windows = 0, directory = 0x0, timestamp_format = 0x0, clientconfdir = 0x0, ssl_dhfile = 0x0, max_children = 0, 
          max_status_children = 0, client_lockdir = 0x0, umask = 18, max_hardlinks = 10000, max_storage_subdirs = 30000, forking = 1, daemon = 1, directory_tree = 1, ca_conf = 0x0, 
          ca_name = 0x0, ca_server_name = 0x0, ca_burp_ca = 0x641330 "/usr/sbin/burp_ca", password_check = 1, manual_delete = 0x0, cname = 0x6412d0 "XXXXXX", 
          password = 0x6412b0 "Y|9Sc2l8|U1RsG", passwd = 0x0, server = 0x641290 "backup1.inikup.com", encryption_password = 0x0, autoupgrade_os = 0x0, autoupgrade_dir = 0x0, 
          ca_csr_dir = 0x641350 "/etc/burp/CA-client", sdcount = 4, startdir = 0x677c80, iecount = 5, incexcdir = 0x66fb70, fscount = 0, fschgdir = 0x0, nbcount = 0, nobackup = 0x0, 
          incount = 0, incext = 0x0, excount = 2, excext = 0x675e10, ircount = 0, increg = 0x0, ercount = 2, excreg = 0x675600, exfscount = 0, excfs = 0x0, excmcount = 2, excom = 0x676190, 
          igcount = 0, incglob = 0x0, cross_all_filesystems = 0, read_all_fifos = 0, fifos = 0x0, ffcount = 0, read_all_blockdevs = 0, blockdevs = 0x0, bdcount = 0, min_file_size = 0, 
          max_file_size = 0, split_vss = 0, strip_vss = 0, vss_drives = 0x0, atime = 0, overwrite = 0, strip = 0, backup = 0x0, restoreprefix = 0x0, regex = 0x0, browsefile = 0x0, 
          browsedir = 0x0, backup_script_pre = 0x6415c0 "/etc/burp/pre_backup_script", backup_script_pre_arg = 0x0, bprecount = 2, backup_script_post = 0x0, backup_script_post_arg = 0x0, 
          bpostcount = 0, backup_script_post_run_on_fail = 0, restore_script_pre = 0x0, restore_script_pre_arg = 0x0, rprecount = 0, restore_script_post = 0x0, 
          restore_script_post_arg = 0x0, rpostcount = 0, restore_script_post_run_on_fail = 0, server_script_pre = 0x0, server_script_pre_arg = 0x0, sprecount = 0, 
          server_script_pre_notify = 0, server_script_post = 0x0, server_script_post_arg = 0x0, spostcount = 0, server_script_post_run_on_fail = 0, server_script_post_notify = 0, 
          backup_script = 0x0, backup_script_arg = 0x0, bscount = 0, restore_script = 0x0, restore_script_arg = 0x0, rscount = 0, server_script = 0x0, server_script_arg = 0x0, sscount = 0, 
          server_script_notify = 0, hardlinked_archive = 0, kpcount = 0, keep = 0x0, working_dir_recovery_method = 0x0, librsync = 1, compression = 9, version_warn = 1, resume_partial = 0, 
          hard_quota = 0, soft_quota = 0, timer_script = 0x0, timer_arg = 0x0, tacount = 0, notify_success_script = 0x0, notify_success_arg = 0x0, nscount = 0, 
          notify_success_warnings_only = 0, notify_success_changes_only = 0, notify_failure_script = 0x0, notify_failure_arg = 0x0, nfcount = 0, rclients = 0x0, rccount = 0, 
          dedup_group = 0x0, client_can_delete = 1, client_can_force_backup = 1, client_can_list = 1, client_can_restore = 1, client_can_verify = 1, server_can_restore = 0, 
          send_client_counters = 1, restore_client = 0x0, restore_path = 0x0, orig_client = 0x0}
        browsefile = 0x0
```

This problem occurs (at least) with Burp 1.3.48 (Debian package) and 1.4.40 on Debian 8.1 (Jessie).

**Comment by grke** [comment]
Thanks, I will try to reproduce and fix this.

**Comment by grke** [comment]
Hello,

I can reproduce this in 1.4.40 and 1.3.48.
It doesn't happen if you are not setting includes/excludes on the server side.
It is already fixed in burp-2.0.x, in which I almost completely re-wrote the conf code.

Unfortunately, having had a go at looking at the burp-1.x.x code to fix it there, I am not
currently able to do so. I suspect that to fix it properly, I would need to backport all the
burp-2.0.x code, which would be pointless as you are better off just moving to burp-2.0.x
instead.

I need to move onto doing things that I consider of a higher priority. Since this has been
broken for a long time already, I am going to leave it alone for the short term.

For now, your choices are to either fix it yourself, or move to burp-2.0.x.

**Comment by jleroy** [comment]
Hi Graham, 

Is Burp 2 suitable for production use ?

**Comment by grke** [comment]
Hello,
Of course, you could just use a wrapper script and put your parameters in there so that you don't need to set them in the burp.conf.

Burp2 is good enough for my personal with protocol 1. Protocol 2 is not ready. You should set protocol=1 in the burp-server.conf.

The thing with burp-2 is that there is a lot more automatic testing in it than there is with burp-1. For example, all the conf loading stuff is automatically tested.
This lets me have confidence that things I change do not break functionality. However, burp-1 has had more 'real world' testing by people other than me.

**Comment by jleroy** [comment]
I've avoided this bug by moving all my includes and excludes directives to client's burp.conf.
Everything works fine now.

Thanks for your help :)

**Comment by grke** [comment]
OK, cool. I will close this issue, as it is fixed in burp-2.0.x.
Thanks for the report.
